Web Components section (#13)
nodeshift/web-application-reference: docs/front-end/web-components.md

# Web Components

The [WHATWG](https://html.spec.whatwg.org/multipage/custom-elements.html) and [WICG](https://github.com/WICG/webcomponents) maintain suite of web technologies and specifications including (but not limited to)

- [Custom Elements](https://html.spec.whatwg.org/multipage/custom-elements.html#custom-elements)
- [Shadow DOM](https://dom.spec.whatwg.org/#shadow-trees)
- The [`<template>` element](https://html.spec.whatwg.org/multipage/scripting.html#the-template-element)
- The [`<slot>` element](https://html.spec.whatwg.org/multipage/scripting.html#the-slot-element)
- [`ElementInternals`](https://html.spec.whatwg.org/multipage/custom-elements.html#element-internals)
- [CSS Custom Properties](https://drafts.csswg.org/css-variables/)
- [Module scripts](https://html.spec.whatwg.org/#module-script)

These are collectively referred to as "web components", and represent the browsers' native component model for web developers.

> Broadly speaking, every web component is a custom element - a specific, globally-recognized tag-name which becomes associated with a class extending HTMLElement in JavaScript. Web components can also make use of the related technologies.

Reviewer: ```suggestion
Broadly speaking, every web component is a custom element - a specific tag-name which becomes associated with a class extending HTMLElement in JavaScript. Web components can also make use of the related technologies.
```

[Scoped Custom Element Registries](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/Scoped-Custom-Element-Registries.md) is almost ready for prime time with [webkit support](https://bugs.webkit.org/show_bug.cgi?id=267454) completed and [gecko](https://bugzilla.mozilla.org/show_bug.cgi?id=1874414) not far behind.

> ## Recommended Components

Reviewer: maybe we should list the [web components community group](https://github.com/w3c/webcomponents-cg/) here?

Author: yea,  that sounds good. i will add it

> - [Lit](https://lit.dev) (for authoring components)

Reviewer: ```suggestion
- [Lit](https://lit.dev) (library and framework for authoring components)
```

#### Microfrontends

This is another case where web components and shadow dom shine, as the encapsulation guarantees and framework interoperability aspects lend themselves naturally to enabling teams to cooperate while maintaining boundaries.

#### Greenfield SPA

There’s an argument to be made from a loading / rendering perf and future-proofing perspective to start new projects with a web component framework (like lit, stencil, hybrids, FAST, etc) in place of traditional / legacy framework (angular, react, vue)

docs.redhat.com (a greenfield project) leaned heavily on RHDS web components, and that helped ship the project faster

#### Piecemeal migration

The framework interoperability aspect enables upgrading complex projects on a component-by-component basis. Rather than break the entire app when bumping the underlying frontend-framework from vX to vY, replace components one by one with interoperable web components.

### Common Gripes

As with all programming models, there are some gotchas that the team has encountered:

- cross-root aria
  - To prevent a11y failures, signifacant expertise might be needed

> - double-registration errors

Reviewer: ```suggestion
- double-registration errors when bundling modules. we recommend using a global import map for each page
```

> - server-side tooling / ssr limitations

Reviewer: ```suggestion
- server-side tooling / ssr limitations. When using the `lit-ssr` library, hydration mismatches when client initial state differs from server rendered state can be difficult to recover from.
```

> - shadow dom can be a double edged sword

Reviewer: ```suggestion
- shadow DOM can be a double edged sword. Teams which are unused to it may find it difficult to work with at first
```

- Unit Testing

>   - JSDOM/Jest may not work with your customer elements

Reviewer: ```suggestion
  - JSDOM/Jest may not work with your custom elements: teams may have to author their components with node environments / DOM shims in mind, which can increase payloads to the client in some cases.
```

### General Concensus

> If you have a team with web platform experience, that always have an MDN tab open, and you’re hopeful to leverage current and future web standards like PWA, then you’re likely to succeed. If on the other hand you don’t have much senior experience or depth of knowledge on web platform features, you may run into issues. However, your teams that overcome those issues will gain platform knowledge and not just framework knowledge.

Reviewer: ```suggestion
Authoring web components: If you have a team with web platform experience, that always have an MDN tab open, and you’re hopeful to leverage current and future web standards like PWA, then you’re likely to succeed. If on the other hand you don’t have much senior experience or depth of knowledge on web platform features, you may run into issues. However, your teams that overcome those issues will gain platform knowledge and not just framework knowledge.

Using web components inside of other major frameworks: our teams have consistently reported that consuming web components inside other frameworks is typically a turnkey developer experience which saves time and reduces duplication of effort.
```

These are collectively referred to as "web components", and represent the browsers' native component model for web developers.

Broadly speaking, every web component is a custom element - a specific tag-name which becomes associated with a class extending HTMLElement in JavaScript. Web components can also make use of the related technologies.

## Recommended Components

- [Lit](https://lit.dev) (library and framework for authoring components)
- [Web Dev Server](https://modern-web.dev) / Web Test Runner (for unit tests and local dev)
- [Custom Elements Manifest](https://github.com/webcomponents/custom-elements-manifest) (for ide support / docgen / codegen)
- [PatternFly Elements](https://patternflyelements.org) Design System
- [Carbon Web Components](https://web-components.carbondesignsystem.com/) Design System
- [Web Components Community Group](https://github.com/w3c/webcomponents-cg/)

## Guidance

### Our Experiences Using Web Components at Red Hat

The team at Red Hat has been involved in several Web Component projects and have learned a good amount about Web Components and how to write them

Two Web Component "Systems" at Red Hat are

> 

Reviewer 2: like the real world examples and use cases here 👍

- [PatternFly Elements](https://patternflyelements.org/) (aka PFE)  (“upstream”)
- [Red Hat Design System](https://ux.redhat.com/) (aka RHDS) (“downstream”)

> ### Advantages and Ideal Use Cases

Reviewer: We should add a section here on architecture:

- state management
- SPA routing
- data fetching

With an emphasis on the concept that web components are agnostic to all those things, and thus can integrate with most or all. For example, there are implementations of redux, mobx, apollo, etc etc for web components

Through the teams experience with Web Components, we've identified a few different pros and cons when using Web Components and what type of applications you might want to use them with.

> 

Reviewer 2: do you think it makes sense to add a bullet about web components being "Framework Agnostic"?

A line about independence of web components in relation to react, vue or angular for example. Not sure if we want to show their usage alongside or independent of these frameworks as well.

The case for web component-based design systems is strong because of encapsulation / knowledge transfer / future proofing as described above. These can also include a certain amount of business logic like analytics code, personalization, etc (ymmv)

#### Microfrontends

This is another case where web components and shadow dom shine, as the encapsulation guarantees and framework interoperability aspects lend themselves naturally to enabling teams to cooperate while maintaining boundaries.

#### Greenfield SPA

There’s an argument to be made from a loading / rendering perf and future-proofing perspective to start new projects with a web component framework (like lit, stencil, hybrids, FAST, etc) in place of traditional / legacy framework (angular, react, vue)

docs.redhat.com (a greenfield project) leaned heavily on RHDS web components, and that helped ship the project faster

#### Piecemeal migration

The framework interoperability aspect enables upgrading complex projects on a component-by-component basis. Rather than break the entire app when bumping the underlying frontend-framework from vX to vY, replace components one by one with interoperable web components.

> ### Common Gripes

Reviewer: we ought to flesh some of these out a bit

As with all programming models, there are some gotchas that the team has encountered:

- cross-root aria
  - To prevent a11y failures, signifacant expertise might be needed

- double-registration errors when bundling modules. we recommend using a global import map for each page

> - server-side tooling / ssr limitations. When using the `lit-ssr` library, hydration mismatches when client initial state differs from server rendered state can be difficult to recover from.

Reviewer: ```suggestion
- <abbr title="Server-side Rendering">SSR</abbr> gotchas. When using the `lit-ssr` library to generate [declarative shadow DOM templates](https://web.dev/articles/declarative-shadow-dom), hydration mismatches when client initial state differs from server rendered state can be difficult to recover from.
```

- shadow DOM can be a double edged sword. Teams which are unused to it may find it difficult to work with at first

- Unit Testing

>   - JSDOM/Jest may not work with your custom elements: teams may have to author their components with node environments / DOM shims in mind, which can increase payloads to the client in some cases.

Reviewer 2: should we include or recommend testing strategies for web components then? Perhaps testing libraries like @open-wc/testing

### General Concensus

Authoring web components: If you have a team with web platform experience, that always have an MDN tab open, and you’re hopeful to leverage current and future web standards like PWA, then you’re likely to succeed. If on the other hand you don’t have much senior experience or depth of knowledge on web platform features, you may run into issues. However, your teams that overcome those issues will gain platform knowledge and not just framework knowledge.

> 

Reviewer 2: not sure if we'd like to emphasize on the performance considerations for web components. 

- There have been experiences when excessive DOM manipulation has resulted in performance bottlenecks. So efficient DOM updates  are crucial for rendering a smooth experience.
- Shadow DOM overhead.
- Event handling overhead.

Reviewer: browser engines are able to optimize rendering based on the encapsulation guarantees, as well

Reviewer: and lit's rendering performance beats react most of the time (scales with interpolations instead of nodes)

Reviewer: we should link out to perf studies on lit and others, and emphasize that b/c wc relies on code already in the browser, it reduces page payloads for end users

Reviewer: https://www.youtube.com/watch?v=CJYkwDTaHzY vaadin lit vs react perf comparison

Reviewer: https://bennypowers.dev/decks/devconf-brno-2023/#slide-27/step-1 toy app page loading comparison (2023 devconf.cz)

react: 2.5mb
lit: 230kb
